src/arithmetic/circuit.rs

```diff
@@ -0,0 +1,118 @@
+use num_bigint::BigUint;
+
+use crate::{
+    circuit::PolyCircuit,
+    gadgets::{
+        crt::{CrtContext, CrtPoly},
+        isolate::IsolationGadget,
+    },
+    poly::Poly,
+};
+use std::sync::Arc;
+
+#[derive(Clone)]
+pub struct ArithmeticCircuit<P: Poly> {
+    pub limb_bit_size: usize,
+    pub num_crt_limbs: usize,
+    pub packed_limbs: usize,
+    pub num_inputs: usize,
+    pub poly_circuit: PolyCircuit<P>,
+    pub ctx: Arc<CrtContext<P>>,
+    pub use_reconstruction: bool,
+    // inputs + intermediate results
+    pub all_values: Vec<CrtPoly<P>>,
+}
+
+impl<P: Poly> ArithmeticCircuit<P> {
+    pub fn to_poly_circuit(&mut self, isolate_gadget: IsolationGadget<P>) {
+        let input_gates = self.poly_circuit.input(self.num_inputs);
+        let mut decomposed_outputs = Vec::with_capacity(input_gates.len() * self.packed_limbs);
+        for g in input_gates {
+            let isolated = isolate_gadget.isolate_terms(&mut self.poly_circuit, g);
+            decomposed_outputs.extend(isolated);
+        }
+        self.poly_circuit.output(decomposed_outputs);
+    }
+
+    pub fn setup(
+        params: &<P as Poly>::Params,
+        num_crt_limbs: usize,
+        limb_bit_size: usize,
+        packed_limbs: usize,
+        inputs: &[BigUint],
+        use_reconstruction: bool,
+    ) -> Self {
+        let num_inputs = inputs.len();
+        let mut poly_circuit = PolyCircuit::<P>::new();
+        let ctx = Arc::new(CrtContext::setup(&mut poly_circuit, &params, limb_bit_size));
+        let mut crt_inputs = Vec::with_capacity(num_inputs);
+        for _ in 0..num_inputs {
+            let crt_poly = CrtPoly::input(ctx.clone(), &mut poly_circuit);
+            crt_inputs.push(crt_poly);
+        }
+
+        ArithmeticCircuit {
+            limb_bit_size,
+            num_crt_limbs,
+            packed_limbs,
+            num_inputs,
+            poly_circuit,
+            ctx,
+            use_reconstruction,
+            all_values: crt_inputs,
+        }
+    }
+
+    /// rhs + lhs
+    pub fn add(&mut self, rhs_index: usize, lhs_index: usize) -> usize {
```

Comment:
Can you define a new struct `ArithGateId` that just wraps `usize` because the usize type is too agnostic to an application context and ambiguous.